### PR TITLE
pass `RUSTC_HOST_FLAGS` at once without the for loop

### PR DESCRIPTION
**Pull request by onur-ozkan**
For obvious reasons...

### CODE REVIEW

File: src/bootstrap/src/utils/shared_helpers.rs [hunk 1]
```diff
@@ -60,16 +64,30 @@ pub fn exe(name: &str, target: &str) -> String {
...
+        // it repeatedly as it’s called thousands of times from the rustc shim.
+        INIT.call_once(|| {
+            VERBOSITY = match env::var("RUSTC_VERBOSE") {
+                Ok(s) => usize::from_str(&s).expect("RUSTC_VERBOSE should be an integer"),
+                Err(_) => 0,
+            };
+        });
+
+        VERBOSITY
     }
```
Comment (fmease) [inline src/bootstrap/src/utils/shared_helpers.rs:84]:
```suggestion
    static VERBOSITY: LazyLock<usize> = LazyLock::new(|| {
        env::var("RUSTC_VERBOSE").map_or(0, |s| {
            s.parse().expect("RUSTC_VERBOSE should be an integer")
        })
    });

    *VERBOSITY
```

Comment (onur-ozkan) [reply src/bootstrap/src/utils/shared_helpers.rs:84]: I intentionally preferred `Once` because I thought it was the cheapest way, am I wrong on this?

Comment (fmease) [reply src/bootstrap/src/utils/shared_helpers.rs:84]: I guess it's slightly cheaper than `LazyLock` as it generates fewer instructions. I've checked Godbolt and run a micro benchmark for this function but couldn't come to a good conclusion wrt. performance difference (likely case: my benchmark sucked).

**Comment by Kobzol** [comment]
Was there any specific motivation for these changes? I'm like 99% sure that this does nothing to speed up rustc invocations (I don't think there's anything to optimize in the shims, tbh), and I don't think that introducing unsafe stuff is warranted here (same with the `#[inline(always)]`). IMO in this case "nothing" is better than "better than nothing" here.

**Comment by onur-ozkan** [comment]
> Was there any specific motivation for these changes? I'm like 99% sure that this does nothing to speed up rustc invocations (I don't think there's anything to optimize in the shims, tbh), and I don't think that introducing unsafe stuff is warranted here (same with the #[inline(always)]). IMO in this case "nothing" is in fact better than "better than nothing".

It doesn't make sense to parse a value thousands of times when it's constant throughout the entire bootstrap invocation. For the same reason, it makes more sense to use `inline(always)` on the called functions. It should impact the time spent before calling the real rustc but like I said, nothing significant.

**Comment by onur-ozkan** [comment]
> I don't think there's anything to optimize in the shims, tbh

I think there are still cases we could consider improving, such as parsing values from the arguments by iterating over them for each value.

**Comment by Kobzol** [comment]
What do you mean by "thousands of times"? The function is called exactly [once](https://github.com/rust-lang/rust/blob/5c0131641d2edfc21e4f627f94f0aadc6cfc6e47/src/bootstrap/src/bin/rustc.rs#L36) per rustc/rustdoc invocation. Caching it in a static variable doesn't help with anything, because after rustc finishes executing, its process ends and the static variable is destroyed, along with everything else in the process. And the next rustc invocation just creates a new static variable.

The same is for the inline annotations, the compiler sees the definition of the functions, and thus doesn't need the inline annotations. It can decide on its own whether to inline or not, and in fact inlining them might not be the best choice, perf. wise. The functions are not particularly small or called often, so the function call overhead is not important But in any case, it won't actually change anything, this is completely inert in regards of rustc performance.

The time spent in the shim is inconsequential when compared to the actually invoked rustc, there's really no point in optimizing any of this.

**Comment by onur-ozkan** [comment]
> What do you mean by "thousands of times"? The function is called exactly [once](https://github.com/rust-lang/rust/blob/5c0131641d2edfc21e4f627f94f0aadc6cfc6e47/src/bootstrap/src/bin/rustc.rs#L36) per rustc/rustdoc invocation. Caching it in a static variable doesn't help with anything, because after rustc finishes executing, its process ends and the static variable is destroyed,

Yeah I missed that point, I was thinking that the cache lives as long as the main bootstrap process for a moment..

> The same is for the inline annotations, the compiler sees the definition of the functions, and thus doesn't need the inline annotations. It can decide on its own whether to inline or not, and in fact inlining them might not be the best choice, perf. wise. The functions are not particularly small or called often, so the function call overhead is not important But in any case, it won't actually change anything, this is completely inert in regards of rustc performance.

The aim is to reduce the wrapper overhead, not to improve rustc's overall performance. If something is called frequently, wouldn’t inlining it make it cheaper than invoking it as a separate function? Doesn’t that save some extra instructions?

**Comment by Kobzol** [comment]
> The aim is to reduce the wrapper overhead, not to improve rustc's overall performance. If something is called frequently, wouldn’t inlining it make it cheaper than invoking it as a separate function? Doesn’t that save some extra instructions?

Well, in theory, yes, but inlining doesn't always help performance. And maybe it was already inlined before. But what I wanted to say is that it doesn't really matter what we do in the shim. Even a `check` build of a hello world app performs more than 35 million instructions in `rustc` (according to rustc-perf). And in bootstrap we usually don't invoke just hello world :) Compiling e.g. `syn` is >35 *billion* instructions. So saving a few instructions to avoid a few function calls per rustc invocation doesn't do anything. What I'm trying to say is that this is literally a drop in the ocean and there is no point in optimizing things like this in the shim - it won't be measurable at all. It would save nanoseconds at best :)

**Comment by onur-ozkan** [comment]
I didn’t expect significant improvements [here](https://github.com/rust-lang/rust/pull/132365#issue-2624696013)—it just made more sense to do (for instance, why iterate over a list when we can pass it directly?). If something can be done in a more sensible way, why not doing it? That's why it’s described as "better than nothing".

I get your point and I already mentioned that "no significant improvement was achieved". Unless the change is harmful to the workflow, I don't see a problem with it.

**Comment by Kobzol** [comment]
Yeah, the `cmd.args` change is actually a small readability improvement :) So I'm fine with that.

I still think that the `#[inline(always)]` annotations are completely useless here, but they also don't really hurt anything (they might raise some eye-brows, but whatever).

So feel free to r=me once CI is green.

**Comment by onur-ozkan** [comment]
> I still think that the #[inline(always)] annotations are completely useless here, but they also don't really hurt anything (they might raise some eye-brows, but whatever).

I am fan of making super micro optimizations, but I don’t want to debate over something that doesn’t bring significant results. For that reason I reverted the inline change.

**Pull request merged**